mlrun/utils/vault.py

```diff
@@ -0,0 +1,269 @@
+# Copyright 2020 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import requests
+import os
+from os.path import expanduser
+import json
+from .helpers import logger
+from ..config import config as mlconf
+from ..k8s_utils import get_k8s_helper
+
+vault_default_prefix = "v1/secret/data"
+vault_url_env_var = "MLRUN_VAULT_URL"
+vault_token_env_var = "MLRUN_VAULT_TOKEN"
+vault_role_env_var = "MLRUN_VAULT_ROLE"
+
+
+class VaultStore:
+    def __init__(self, token=None):
+        self._token = token
+        self.url = mlconf.secret_stores.vault.url or os.environ.get(vault_url_env_var)
+
+    @property
+    def token(self):
+        if not self._token:
+            self._login()
+
+        return self._token
+
+    def _login(self):
+        if self._token:
+            return
+
+        self._token = os.environ.get(vault_token_env_var)
+        if self._token:
+            return
+
+        vault_role = None
+        role_env = os.environ.get(vault_role_env_var)
+        if role_env:
+            role_type, role_val = role_env.split(":", 1)
+            vault_role = "mlrun-role-{}-{}".format(role_type, role_val)
+
+        self._login_with_jwt_token(vault_role)
+        if self._token is None:
+            logger.warning(
+                "Vault login: no vault token is available. No secrets will be accessible"
+            )
+
+    @staticmethod
+    def _generate_vault_path(
+        prefix=vault_default_prefix,
+        user=None,
+        project=None,
+        user_prefix="users",
+        project_prefix="projects",
+    ):
+        full_path = prefix + "/mlrun/{}/{}"
+        if user:
+            return full_path.format(user_prefix, user)
+        elif project:
+            return full_path.format(project_prefix, project)
```

Comment:
I suggest to raise an error when both user and project are given

Reply:
Sure, why not.